### ISSUE TITLE
error when using generateId (wrong number of argument)

### ISSUE THREAD

**Issue by scauglog** [logstash-plugins/logstash-output-mongodb#25]
When i use the generateId option, it raise an exception : 

```
wrong number of arguments calling `initialize` (2 for 0)
/opt/logstash-2.2.0/vendor/bundle/jruby/1.9/gems/logstash-output-mongodb-2.0.3/lib/logstash/outputs/mongodb.rb:54
```

according to doc http://www.rubydoc.info/github/mongodb/bson-ruby/BSON/ObjectId#from_string-class_method
you should replace mongodb.rb line 54 

``` ruby
document["_id"] = BSON::ObjectId.new(nil, event["@timestamp"])
```

by something like 

``` ruby
document["_id"] =  BSON::ObjectId.from_time(event["@timestamp"])
```

with `event["@timestamp"]` a Time object or timestamp

**Comment by hhj1988** [comment]
hello ,Have you solved the problem?

**Comment by scauglog** [comment]
I don't have access to this code anymore but I think I use the ruby code filter plugin to set the "_id" with the fix I propose or I generate a random uuid with ruby securerandom.

**Comment by hhj1988** [comment]
I need'_id'to customize the content, the official website says: If true, an "_id" field will be added to the document before insertion. The "_id" field will use the timestamp of the event and overwrite an existing "_id" field in the event.

To configure
input {
stdin{}
}

output{

mongodb {
collection => "std"
database => "fragment"
generateId => true
uri => "mongodb://192.168.199.7:27017"
}
}
~

Failed to send event to MongoDB {:event=>#<LogStash::Event:0x2ef73b32 @metadata={}, @accessors=#<LogStash::Util::Accessors:0x694f67b4 @store={"message"=>"hello", "@version"=>"1", "@timestamp"=>"2016-09-02T05:33:49.575Z", "host"=>"mini03"}, @lut={"host"=>[{"message"=>"hello", "@version"=>"1", "@timestamp"=>"2016-09-02T05:33:49.575Z", "host"=>"mini03"}, "host"], "@timestamp"=>[{"message"=>"hello", "@version"=>"1", "@timestamp"=>"2016-09-02T05:33:49.575Z", "host"=>"mini03"}, "@timestamp"]}>, @data={"message"=>"hello", "@version"=>"1", "@timestamp"=>"2016-09-02T05:33:49.575Z", "host"=>"mini03"}, @metadata_accessors=#<LogStash::Util::Accessors:0x1f036be0 @store={}, @lut={}>, @cancelled=false>, :exception=>#, :backtrace=>["/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-output-mongodb-2.0.5/lib/logstash/outputs/mongodb.rb:54:in `receive'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/outputs/base.rb:109:in`multi_receive'", "org/jruby/RubyArray.java:1613:in `each'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/outputs/base.rb:109:in`multi_receive'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/output_delegator.rb:130:in `worker_multi_receive'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/output_delegator.rb:114:in`multi_receive'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:301:in `output_batch'", "org/jruby/RubyHash.java:1342:in`each'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:301:in `output_batch'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:232:in`worker_loop'", "/export/servers/logstash-2.4.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.0-java/lib/logstash/pipeline.rb:201:in `start_workers'"], :level=>:warn}

THK!

**Comment by scauglog** [comment]
If you just need unique id don't use the generateId option. If there is no _id in your event mongo will automatically generate a new id for each record you send.

**Comment by hhj1988** [comment]
I need "_id" field will use the timestamp of the event and overwrite an existing "_id" field in the event.  configure  generateId ,Am I still missing?

**Comment by scauglog** [comment]
then fork or clone the repo, do the fix I propose, [build the gem](https://github.com/logstash-plugins/logstash-output-mongodb#22-run-in-an-installed-logstash). [uninstall the mongo plugin](https://www.elastic.co/guide/en/logstash/current/working-with-plugins.html#removing-plugins) and [install a locally built plugin](https://www.elastic.co/guide/en/logstash/current/working-with-plugins.html#installing-local-plugins)

**Comment by hhj1988** [comment]
THK！but,I do not know how to operate, can you say more details.

**Comment by scauglog** [comment]
before doing anything you should probably try with generateId set to false and see if the result could work for your use case.

``` sh
cd ~
git clone https://github.com/logstash-plugins/logstash-output-mongodb.git
cd logstash-output-mongodb
```

edit lib/logstash/mongodb.rb file line 54 

``` sh
gem build logstash-output-mongodb.gemspec
cd to/your/logstash
bin/logstash-plugin uninstall logstash-output-mongodb
bin/logstash-plugin install ~/logstash-output-mongodb/logstash-output-mongodb-2.0.5.gem
```

**Comment by hhj1988** [comment]
THK！Installation problems。
gem build logstash-output-mongodb.gemspec?
ERROR:  While executing gem ... (NoMethodError)
    undefined method `metadata=' for #Gem::Specification:0x7f4dfb8da400

**Comment by colinsurprenant** [comment]
This should be solved by #57 
Let me know if you think there is a good reason to actually use `BSON::ObjectId.from_time` instead of a bare `BSON::ObjectId.new` as proposed in that PR.
